bpo-30406: Make async and await proper keywords

JelleZijlstra opened the pull request:
<!-- issue-number: bpo-30406 -->
https://bugs.python.org/issue30406
<!-- /issue-number -->

JelleZijlstra committed:
make async and await real keywords

JelleZijlstra committed:
remove asyncio test that is now a SyntaxError

JelleZijlstra committed:
update syntax tests in test_coroutines

JelleZijlstra committed:
regenerate symbol.py

JelleZijlstra committed:
fix tokenize.py

JelleZijlstra committed:
fewer nested lists allowed

Maybe because I added await_expr

JelleZijlstra committed:
regenerate keyword.py

JelleZijlstra committed:
fix lib2to3

JelleZijlstra committed:
add some docs

AraHaan reviewed (approved):
Looks alright to me.

JelleZijlstra commented:
I might need some time to fix the test failures.

AraHaan commented:
Alright, But in general it looks like a nice start.

JelleZijlstra committed:
fix test_parser.py

JelleZijlstra committed:
attempt to get pydoc to work

vstinner commented on Lib/lib2to3/tests/test_parser.py:
```diff
@@ -167,34 +167,34 @@ def foo(): pass
-        self.invalid_syntax("""def foo():
-                                   async for a in b: pass""")
+        self.validate("""def foo():
+                             async for a in b: pass""")
```
wait... this one looks like a bug.

JelleZijlstra replied:
I think lib2to3's parser in general doesn't do semantic checking of the sort that would be required to reject this. For example, it also doesn't fail on putting `nonlocal` or `yield from` outside a function, or `yield from` inside an async generator.

vstinner commented on Lib/lib2to3/tests/test_parser.py:
```diff
@@ -167,34 +167,34 @@ def foo(): pass
-        self.invalid_syntax("""def foo():
-                                   async with a: pass""")
+        self.validate("""def foo():
+                             async with a: pass""")
```
ditto (also a bug)

JelleZijlstra replied:
Same here, this is also lib2to3.

1st1 commented:
Overall looks good. I'll make a detailed review in a few days.

vstinner reviewed (commented):
https://docs.python.org/dev/reference/grammar.html should also be updated, no?

JelleZijlstra commented:
grammar.html just directly includes Grammar/Grammar, so it shouldn't require further changes.

AraHaan commented:
@JelleZijlstra you mind also removing 'asyncio.async' as well as that thing might cause an SyntaxError.

vstinner commented:
Thanks for the clarification :-) I let Yury review the change. He knows the
parser much better than me!

JelleZijlstra commented:
@AraHaan it's currently implemented as something like `globals()['async'] = ensure_future`, so the implementation won't cause a SyntaxError (as you can tell from the test suite passing). I suppose I could remove the alias, but I also don't want to make this patch bigger than necessary, so I'll wait for Yury to review it.

AraHaan commented:
Ok

1st1 commented:
I'll merge this when I return from my vacation in 11 days. Thank you for keeping the pr up to date!

JelleZijlstra commented:
@1st1 ping – have you had a chance to look at this PR?

1st1 reviewed (commented):
Overall looks good. Could you please rename it once again and add a news entry with blurb?

1st1 commented:
s/rename/rebase

JelleZijlstra committed:
add blurb

JelleZijlstra commented:
Just rebased from current master and added a news entry using blurb.

JelleZijlstra committed:
run make patchcheck to fix whitespace

JelleZijlstra commented:
The AppVeyor failure is in test_idle and looks unrelated to this PR.

JelleZijlstra commented:
Oh, probably related to 7c5798ebfa9b822d6cedf1fc8c034f6ea90644dc.

vstinner commented:
> The AppVeyor failure is in test_idle and looks unrelated to this PR.

Rebase again and it should be fine.

1st1 commented on Grammar/Grammar:
```diff
@@ -133,13 +134,15 @@ arglist: argument (',' argument)*  [',']
+comp_iter: comp_async_for | comp_for | comp_if
+comp_async_for: 'async' 'for' exprlist 'in' or_test [comp_iter]
+comp_for: 'for' exprlist 'in' or_test [comp_iter]
 comp_if: 'if' test_nocond [comp_iter]
```
Why did you do all these renames and added `atom_expr`? Why don't we just replace `ASYNC` with `'async'` (ditto for await)?  I'd avoid making **any** changes to the grammar besides token adjustments.

JelleZijlstra replied:
I don't remember exactly (it's been a few months), but I just tried to implement your suggestions and I'm getting failures in test_parser that I don't currently understand. I'll look into it some more. The errors are all on comprehensions and are similar to

```
  File "/Users/jzijlstra-mpbt/py/cpython/Lib/test/test_parser.py", line 24, in roundtrip
    self.fail("could not roundtrip %r: %s" % (s, why))
AssertionError: could not roundtrip '{x for x in seq}': Illegal terminal: expected NAME, got 330.
```

1st1 commented on Lib/lib2to3/Grammar.txt:
```diff
@@ -155,13 +156,15 @@ arglist: argument (',' argument)* [',']
+comp_iter: comp_async_for | comp_for | comp_if
+comp_async_for: 'async' 'for' exprlist 'in' testlist_safe [comp_iter]
+comp_for: 'for' exprlist 'in' testlist_safe [comp_iter]
 comp_if: 'if' old_test [comp_iter]
```
Adding to the point raised for changes in `Grammar`: when you change/rename/add new productions here, you might be breaking some code that uses `lib2to3` to define custom fixters. Again, I'd avoid any changes.

1st1 commented on Lib/test/test_coroutines.py:
```diff
@@ -394,18 +394,14 @@ def async(): pass
 
-    def test_goodsyntax_1(self):
+    def test_badsyntax_4(self):
         # Tests for issue 24619
```
You can drop this comment now.

1st1 commented on Misc/NEWS.d/next/Core and Builtins/2017-07-20-22-03-44.bpo-30406._kr47t.rst:
```diff
@@ -0,0 +1 @@
+``async`` and ``await`` are now proper keywords, as specified in PEP 492.
```
Change to:

```
Make ``async`` and ``await`` proper keywords, as specified in PEP 492.
```

1st1 reviewed (changes requested):
Playing with the code right now. One of the things you also need to do is to cleanup `forbidden_name()` function in `Python/ast.c`.

JelleZijlstra committed:
remove obsolete code from forbidden_name

JelleZijlstra committed:
attempt to make it work without await_expr

JelleZijlstra committed:
undo async_for-related changes

test_parser still fails for some reason

JelleZijlstra committed:
minor CR fixes

JelleZijlstra committed:
undo unnecessary change

1st1 commented:
Ping? I plan to work on this myself if this isn't resolved soon.

JelleZijlstra commented:
I was unable to resolve the issue I mentioned in https://github.com/python/cpython/pull/1669#discussion_r128901454. I can push out the other fixes though and take another look within the next few days to see if I can get it working.

1st1 commented:
> I was unable to resolve the issue I mentioned in #1669 (comment). [..]

Interesting. I'll take a look at that too.

1st1 commented:
So here's my attempt at making this work without extensive grammar changes: https://github.com/1st1/cpython/commit/0eee314f87607fb7cf5178cd0998a9825070af18

If you are interested in continuing to work on this PR, I suggest you to incorporate my approach.

JelleZijlstra commented:
Thanks! I'll spend some time on it tonight (California time).

JelleZijlstra committed:
closer to yuri's code

JelleZijlstra commented:
My attempt was very similar to yours, but your code also fails test_parser (at least when I compile it locally). I'll push what I have for now.

JelleZijlstra committed:
fix whitespace (make patchcheck)

JelleZijlstra committed:
fix test_symbol

1st1 commented:
> My attempt was very similar to yours, but your code also fails test_parser (at least when I compile it locally).

I didn't run the tests, just checked if it was possible to fix the grammar in the minimal way. Some tests should probably be adjusted.

Please also make sure that lib2to3/grammar is in sync with grammar/grammar.

JelleZijlstra commented:
I fixed lib2to3 and adjusted all tests that were obvious.

Unfortunately the `test_parser` problem doesn't look like it's just a matter of adjusting the test. The failures look like this:
```
======================================================================
FAIL: test_set_comprehensions (test.test_parser.RoundtripLegalSyntaxTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_parser.py", line 22, in roundtrip
    st2 = parser.sequence2st(t)
parser.ParserError: Illegal terminal: expected NAME.
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_parser.py", line 400, in test_set_comprehensions
    self.check_expr('{x for x in seq}')
  File "/home/travis/build/python/cpython/Lib/test/test_parser.py", line 30, in check_expr
    self.roundtrip(parser.expr, s)
  File "/home/travis/build/python/cpython/Lib/test/test_parser.py", line 24, in roundtrip
    self.fail("could not roundtrip %r: %s" % (s, why))
AssertionError: could not roundtrip '{x for x in seq}': Illegal terminal: expected NAME.
```
I couldn't figure out why this doesn't work when all other tests pass. Something must be wrong related to the code for async comprehensions but I'm not sure what.

1st1 commented:
> Unfortunately the test_parser problem doesn't look like it's just a matter of adjusting the test. The failures look like this:

OK, I think I figured this out.  The problem is that when we have `comp_for` defined as follows:

```
comp_for: ['async'] 'for' exprlist 'in' or_test [comp_iter]
```

we have two states in the grammar for `comp_for`:

```
state1: NAME NAME exprlist NAME or_test ...
state2: NAME exprlist NAME or_test ...
```

and the code in `parsermodule.c` (`validate_node()` function) doesn't know how to pick the correct state from all possible states in the generated grammar.  The code in `ast.c` doesn't have this problem as we hardcode states traversal logic.  Theoretically we could fix `validate_node()`, but I'm afraid it would complicate things too much, so let's just add another non-term to the grammar.

I propose to split `comp_for` into two non-terms: `sync_comp_for` and `comp_for`:

```
sync_comp_for: 'for' exprlist 'in' or_test [comp_iter]
comp_for: ['async'] sync_comp_for
```

This seems to be less of a change compared to the first approach in this PR (adding `comp_for_or_async` and `comp_async_for`).

I tried it here https://github.com/1st1/cpython/commit/3457053825d6f96662b816514be11e126cc1014b and it seems to work for all `test_parser` tests.

JelleZijlstra commented:
Thanks for looking into this! The suggested change makes sense. I'll implement that in the next few days and make sure all tests pass (might not have time today).

1st1 commented:
> I'll implement that in the next few days and make sure all tests pass (might not have time today).

No rush, I won't be able to further review till next Monday anyways.

JelleZijlstra committed:
use a separate sync_comp_for (suggested by Yury)

JelleZijlstra committed:
regenerate symbol.py

JelleZijlstra committed:
simplify lib2to3 changes

JelleZijlstra commented:
Just pushed an update that passes all tests locally. I'll check in in a while to make sure Travis passes.

1st1 commented on Doc/library/token.rst:
```diff
@@ -129,9 +127,11 @@ the :mod:`tokenize` module.
    Added :data:`COMMENT`, :data:`NL` and :data:`ENCODING` tokens.
+
+.. versionchanged:: 3.7
+   Removed :data:`AWAIT` and :data:`ASYNC` tokens again. "async" and
```
Drop "again".

JelleZijlstra replied:
Done.

1st1 commented on Python/ast.c:
```diff
@@ -1850,19 +1837,27 @@ ast_for_comprehension(struct compiling *c, const node *n)
+        if (NCH(n) == 2) {
             is_async = 1;
+            REQ(CHILD(n, 0), NAME);
+            assert(strcmp(STR(CHILD(n, 0)), "async") == 0);
```
I think we should add another macro:

```C
#define REQ_NAME(node, name) \
    REQ(node, NAME); \
    assert(strcmp(STR(node), name) == 0);
```

so that 

```C
REQ(CHILD(n, 0), NAME);
assert(strcmp(STR(CHILD(n, 0)), "async") == 0);
```

becomes

```C
REQ_NAME(CHILD(n, 0), "async");
```

JelleZijlstra replied:
Should that go into Include/node.h? I can add it but I'm not sure it deserves to be in the public API, and it seems out of scope for this PR.

1st1 replied:
IIRC "node.h" isn't part of the public API at all. But it's fine, let's skip this bit.

JelleZijlstra replied:
OK, I can make another PR after this one is merged.

1st1 commented:
Looks good. @haypo I'll merge this unless you have more comments.

JelleZijlstra committed:
remove "again" from wording intoken.rst

vstinner commented on Doc/library/token.rst:
```diff
@@ -129,9 +127,11 @@ the :mod:`tokenize` module.
+
+.. versionchanged:: 3.7
+   Removed :data:`AWAIT` and :data:`ASYNC` tokens. "async" and "await" are
+   now tokenized as :data:`NAME` tokens.
```
Why not saying "keywords". These names are added to keywords.py and it's the PR title, no?

vstinner reviewed (approved):
LGTM with a minor question on the doc.

1st1 commented:
@JelleZijlstra Could you please rebase against current master?  I resolved one conflict in lib2to3/Grammar, but there're few other things that broke CI.

JelleZijlstra committed:
fix lib2to3

JelleZijlstra commented:
Just rebased and adjusted the lib2to3 grammar changes to make local tests pass.
